File: src/BaseHmiTypes/Scripts/HmiScriptCommandDefinition.cs
using BaseHmiTypes.Screens.Base;

namespace BaseHmiTypes.Scripts;

public sealed class HmiScriptCommandDefinition
{
    public string Key { get; set; } = string.Empty;

    public string Name { get; set; } = string.Empty;

    public string DisplayName { get; set; } = string.Empty;

    public string? Description { get; set; }

    public string? Category { get; set; }

    public HmiScriptLanguage Language { get; set; } = HmiScriptLanguage.Unknown;

    public Type CommandType { get; set; } = typeof(HmiScriptCommand);

    public IList<HmiScriptCommandParameterDefinition> Parameters { get; } = new List<HmiScriptCommandParameterDefinition>();
}

public sealed class HmiScriptCommandParameterDefinition
{
    public string Name { get; set; } = string.Empty;

    public string PropertyName { get; set; } = string.Empty;

    public string? Description { get; set; }

    public HmiScriptCommandParameterDirection Direction { get; set; } = HmiScriptCommandParameterDirection.In;

    public IList<HmiScriptCommandParameterTypeDefinition> AcceptedTypes { get; } = new List<HmiScriptCommandParameterTypeDefinition>();
}

public sealed class HmiScriptCommandParameterTypeDefinition
{
    public string ObjectType { get; set; } = string.Empty;

    public string? Description { get; set; }

    public string? InitialValue { get; set; }

    public IList<string> ValueTypes { get; } = new List<string>();

    public IList<HmiScriptCommandSelectionEntryDefinition> SelectionEntries { get; } = new List<HmiScriptCommandSelectionEntryDefinition>();
}

public sealed class HmiScriptCommandSelectionEntryDefinition
{
    public string Value { get; set; } = string.Empty;

    public string? Description { get; set; }

    public string? ScriptConstant { get; set; }
}

public enum HmiScriptCommandParameterDirection
{
    In,
    Out,
    Return
}

File: typescript/base-hmi-types/src/scripts/HmiScriptCommandDefinition.ts
import { HmiScriptLanguage } from "../screens/base/HmiProperty.js";
import { HmiScriptCommand, HmiScriptCommandParameterDirection } from "./HmiScriptCommand.js";

export type HmiScriptCommandConstructor = new () => HmiScriptCommand;

export class HmiScriptCommandDefinition {
  key = "";
  name = "";
  displayName = "";
  description?: string;
  category?: string;
  language: HmiScriptLanguage = HmiScriptLanguage.Unknown;
  commandType?: HmiScriptCommandConstructor;
  readonly parameters: HmiScriptCommandParameterDefinition[] = [];
}

export class HmiScriptCommandParameterDefinition {
  name = "";
  propertyName = "";
  description?: string;
  direction: HmiScriptCommandParameterDirection = HmiScriptCommandParameterDirection.In;
  readonly acceptedTypes: HmiScriptCommandParameterTypeDefinition[] = [];
}

export class HmiScriptCommandParameterTypeDefinition {
  objectType = "";
  description?: string;
  initialValue?: string;
  readonly valueTypes: string[] = [];
  readonly selectionEntries: HmiScriptCommandSelectionEntryDefinition[] = [];
}

export class HmiScriptCommandSelectionEntryDefinition {
  value = "";
  description?: string;
  scriptConstant?: string;
}
